packer/core.go

```diff
@@ -386,15 +398,15 @@ func (c *Core) init() error {
 	// We need to read the keys from both, then loop over all of them to figure
 	// out the appropriate interpolations.
 
-	allVariables := make(map[string]string)
+	repeatMap := make(map[string]string)
```

Comment:
I'm not sure wether this is a good idea here, but since ranging on maps is random may be it could be a good idea to put the key/values in a sorted array. The outcome is that an error case will either always or never be reproducible; which is handy, for debugging and testing and all.


```go
// keyValue could be sorted by first checking 
// wether it needs to be interpolated then by key.
type keyValue struct {
  Key   string
  Value string
}

sort.Slice(keyValues, func(i, j int) bool {
  if isDoneInterpolating(kv[i]) && !isDoneInterpolating(kv[j]) {
      return true // put the not interpolated vars first.
  }
  return kv[i].Key < kv[j].Key
})
```

Reply:
Then for each pass we can sort this array

Reply:
I like @azr idea.
You can also do like I did for HCL variables, ranging over a slice of keys and move the none interpolated ones to the end of the slice so they can be interpolated after all other vars. Of course, use the repeatMap to get the key -> value.

Reply:
This is a good idea from a reproducibility perspective.

Reply:
Sorting by whether it's done interpolating seems like overkill. I did a string sort so that it'll be deterministic, but I think that's sufficient.

Reply:
got it 👍